packages/flutter/lib/src/rendering/custom_paint.dart

```diff
@@ -0,0 +1,830 @@
+      _setRasterCacheHints(context);
+    }
+    super.paint(context, offset);
+    if (_foregroundPainter != null) {
+      _paintWithPainter(context.canvas, offset, _foregroundPainter);
+      _setRasterCacheHints(context);
+    }
+  }
+
+  void _setRasterCacheHints(PaintingContext context) {
+    if (isComplex)
+      context.setIsComplexHint();
+    if (willChange)
+      context.setWillChangeHint();
+  }
+
+  /// Builds semantics for the picture drawn by [painter].
+  SemanticsBuilderCallback _backgroundSemanticsBuilder;
+
+  /// Builds semantics for the picture drawn by [foregroundPainter].
+  SemanticsBuilderCallback _foregroundSemanticsBuilder;
+
+  @override
+  void describeSemanticsConfiguration(SemanticsConfiguration config) {
+    super.describeSemanticsConfiguration(config);
+    _backgroundSemanticsBuilder = painter?.semanticsBuilder;
+    _foregroundSemanticsBuilder = foregroundPainter?.semanticsBuilder;
+    config.isSemanticBoundary = _backgroundSemanticsBuilder != null || _foregroundSemanticsBuilder != null;
+  }
+
+  /// Describe the semantics of the picture painted by the [painter].
+  List<SemanticsNode> _backgroundSemanticsNodes;
+
+  /// Describe the semantics of the picture painted by the [foregroundPainter].
+  List<SemanticsNode> _foregroundSemanticsNodes;
+
+  @override
+  void assembleSemanticsNode(
+    SemanticsNode node,
+    SemanticsConfiguration config,
+    Iterable<SemanticsNode> children,
+  ) {
+    assert(() {
+      if (child == null && children.isNotEmpty) {
+        throw new FlutterError(
+          '$runtimeType does not have a child widget but received a non-empty list of child SemanticsNode:\n'
+          '${children.join('\n')}'
+        );
+      }
+      return true;
+    }());
+
+    final List<CustomPainterSemantics> backgroundSemantics = _backgroundSemanticsBuilder != null
+      ? _backgroundSemanticsBuilder(size)
+      : const <CustomPainterSemantics>[];
+    _backgroundSemanticsNodes = _updateSemanticsChildren(_backgroundSemanticsNodes, backgroundSemantics);
+
+    final List<CustomPainterSemantics> foregroundSemantics = _foregroundSemanticsBuilder != null
+      ? _foregroundSemanticsBuilder(size)
+      : const <CustomPainterSemantics>[];
+    _foregroundSemanticsNodes = _updateSemanticsChildren(_foregroundSemanticsNodes, foregroundSemantics);
+
+    final bool hasBackgroundSemantics = _backgroundSemanticsNodes != null && _backgroundSemanticsNodes.isNotEmpty;
+    final bool hasForegroundSemantics = _foregroundSemanticsNodes != null && _foregroundSemanticsNodes.isNotEmpty;
+    final List<SemanticsNode> finalChildren = <SemanticsNode>[];
+    if (hasBackgroundSemantics)
+      finalChildren.addAll(_backgroundSemanticsNodes);
+    finalChildren.addAll(children);
+    if (hasForegroundSemantics)
+      finalChildren.addAll(_foregroundSemanticsNodes);
+    super.assembleSemanticsNode(node, config, finalChildren);
+  }
+
+  /// Updates `semanticsChildren` from `newSemantics`.
```

Comment:
What are `semanticsChildren` and `newSemantics`?

Reply:
Forgot to rename these. Fixed.